code/modules/mob/living/simple_animal/hostile/hostile.dm

```diff
@@ -577,3 +592,57 @@ mob/living/simple_animal/hostile/proc/DestroySurroundings() // for use with mega
 		friends = fren
 		faction = fren.faction.Copy()
 	return ..()
+
+/**
+  * Proc that handles a charge attack for a mob.
+  */
+/mob/living/simple_animal/hostile/proc/enter_charge(var/atom/target)
+	if((mobility_flags & (MOBILITY_MOVE | MOBILITY_STAND)) != (MOBILITY_MOVE | MOBILITY_STAND) || charge_state)
+		return FALSE
+
+	if((charge_cooldown > world.time) || !has_gravity() || !target.has_gravity())
+		return FALSE
+	Shake(15, 15, 1 SECONDS)
+	sleep(1.5 SECONDS) //Provides a visable wind up and tell for all charging mobs, with consistant visuals each time.
+	charge_state = TRUE
+	weather_immunities += "lava"
+	throw_at(target, charge_distance, 1, src, FALSE, TRUE, callback = CALLBACK(src, .proc/charge_end))
+	return TRUE
+
+/mob/living/simple_animal/hostile/proc/charge_end()
+	charge_state = FALSE
+	weather_immunities -= "lava"
```

Comment:
so if any charging mob in the future is lavaproof, it'll lose its lava immunity after its first charge?